Add Instant deposit button in Simplify deposit UI.

shendy-a8c opened the pull request:
Fixes https://github.com/Automattic/woocommerce-payments/issues/5889.

#### Changes proposed in this Pull Request

Show `InstantDepositButton` on the new deposit UI. That button is currently used in the current UI.

NOTE:
- In the new UI, the button text is `Deposit available funds`. Currently, the button says `Instant deposit`. I change the button text here, so when the feature flag `_wcpay_feature_simplify_deposits_ui` is off, the instant deposit button in the current (old) UI will also show the new text. I think it's OK instead of overengineer the button to support different text.
- Same with `isSecondary` and `isPrimary` attribute for that button. The current (old) UI uses `isSecondary` but the new UI uses `isPrimary`.

#### Testing instructions

To get instant deposit:
* Make sure to use USD as merchant currency.
* Add debit card as payout method. Read https://woocommerce.com/document/woocommerce-payments/deposits/instant-deposits/#add-debit-card.
* Purchase so you have balance > USD 50 using `4000000000000077` which will bypass pending balance ([ref](https://stripe.com/docs/testing#available-balance)).
* Enable feature flag `update_option( '_wcpay_feature_simplify_deposits_ui', '1' );` via [WP Console](https://wordpress.org/plugins/wp-console/).
* Open Payments > Overview page.
* A `Deposit available funds` button should show under the balance boxes.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

shendy-a8c committed:
Add Instant deposit button.

shendy-a8c committed:
Changelog.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6004` or branch name `add/5889-instant-deposit` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=add/5889-instant-deposit) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 52f1346d0df9f9e93efc588ff70fcdacdaa1bbec
- Build time: 2023-04-16 21:50:13 UTC

_Note: the build is updated when a new commit is pushed to this PR._

shendy-a8c committed:
Merge branch 'develop' into add/5889-instant-deposit

shendy-a8c committed:
When instant deposit amount is 0, don't show instant deposit section.

shendy-a8c committed:
Update unit test and snapshot.

shendy-a8c committed:
Unit test for instant deposit button on balance panel.

shendy-a8c committed:
Fix unit test.

shendy-a8c committed:
Merge branch 'develop' into add/5889-instant-deposit

shendy-a8c committed:
Update snapshot.

shendy-a8c committed:
Merge branch 'develop' into add/5889-instant-deposit

Jinksi commented on client/components/account-balances/balances-tab-panel.tsx:
```diff
@@ -35,6 +36,7 @@ type BalanceTab = {
 	availableFunds: number;
 	pendingFunds: number;
 	delayDays: number;
+	instant?: AccountOverview.InstantBalance;
```
Nit: I'd suggest naming this `instantBalance` for clarity. I thought `instant` represented a `boolean` at first glance.

shendy-a8c replied:
Good point. Addressed in [7115c26](https://github.com/Automattic/woocommerce-payments/pull/6004/commits/7115c26263b2c819af8cdd4f8cdb3110b830ac3e) and [1507df5](https://github.com/Automattic/woocommerce-payments/pull/6004/commits/1507df5e92aa329adfc60b343134e67383f68445).

Jinksi commented:
**Update:** I've fixed this behaviour in #6019 

As @shendy-a8c mentioned in a slack discussion (p1681297620527469-slack-C045PC0V40G), there is a bug that causes the Instant Deposit modal to disappear (if open) when the deposits history table finishes loading.

https://user-images.githubusercontent.com/3147296/231612646-cc138677-8510-4007-a5f8-72ceb55373ae.mov




This is probably due to the "forced re-render" workaround I used to get the currency tab switcher to select the tab from the URL query param once the data was fetched. (TabPanel is an uncontrolled component, we can only pass in `initialTabName`).

https://github.com/Automattic/woocommerce-payments/blob/58d18e7c213e5ef15be9ba6d1ec16f022f3c4fef/client/components/account-balances/balances-tab-panel.tsx#L95-L98

It wasn’t a problem at the time but is now that there is a child modal that should persist.


I’d suggest a more robust `isLoading` component for the `AccountBalancesTabPanel` component, which will bypass the reason for my “workaround” and make the code for this component’s loading state easier to follow.

If the account overview `isLoading` (we don’t have a list of the account’s currencies yet), return early with a new dedicated loading state component for `AccountBalancesTabPanel`, (e.g. `AccountBalancesTabPanelLoading`).

shendy-a8c committed:
Merge branch 'develop' into add/5889-instant-deposit

shendy-a8c committed:
Change 'instant' to 'instantBalance' to make it more clear.

shendy-a8c committed:
Change 'instant' to 'instantBalance' to make it more clear.

shendy-a8c commented on client/components/account-balances/balances-tab-panel.tsx:
```diff
@@ -101,97 +105,109 @@ const AccountBalancesTabPanel: React.FC = () => {
+								}
+							/>
+						</Flex>
+						{ tab.instantBalance && tab.instantBalance.amount > 0 && (
```
This big diff is just wrapping the above `<Flex>` with `<></>` and adding a conditional instant deposit button after that `</Flex>`.

Jinksi committed:
Merge branch 'develop' into add/5889-instant-deposit

Jinksi commented on client/components/account-balances/test/index.test.tsx:
```diff
@@ -438,4 +450,34 @@ describe( 'AccountBalancesTabPanel', () => {
+		const instantDepositButton = screen.getByRole( 'button', {
+			name: 'Deposit available funds',
+		} );
+		expect( instantDepositButton ).toBeInTheDocument();
```
Nit: since we're using a `get` react-resting-library query, we don't need to also `expect()`, since `getByRole()` represents the same assertion as `expect( instantDepositButton ).toBeInTheDocument()` ([docs ref](https://testing-library.com/docs/queries/about#types-of-queries)).
```suggestion
		screen.getByRole( 'button', {
			name: 'Deposit available funds',
		} );
```

shendy-a8c replied:
Nice! Suggestion committed.

Jinksi committed:
Merge branch 'develop' into add/5889-instant-deposit

Jinksi reviewed (dismissed):
@shendy-a8c this works well with the `_wcpay_feature_simplify_deposits_ui` flag enabled. The re-rendering fix from #6019 is ensuring the modal does not disappear unexpectedly.

I'm not as confident as you in the decision to change the current design before we enable the flag. This may add confusion for our users, since the button will not match our public docs. This will only be a concern if the flag is not flipped in the same WCPay release. Note: this feature is still works correctly, regardless of the design change.

Correct me if I'm wrong, but I don't think the effort would be too high, perhaps a ternary statement in the `<InstantDepositButton />` component. 

**Changes to current design:**

![Instant deposits previous design](https://user-images.githubusercontent.com/3147296/231937206-1e494ab7-2ee8-44b3-a6e4-2de502a7ebb3.png)

**Documentation screenshot ([src](https://woocommerce.com/document/woocommerce-payments/deposits/instant-deposits/)):**

<img width="400" alt="image" src="https://user-images.githubusercontent.com/3147296/231938266-e9479239-c7cc-4291-9064-ce3824019532.png">


If we do change the current design, I think a changelog entry should be added, since it will be a user-facing change with or without the feature flag.

shendy-a8c committed:
Merge branch 'develop' into add/5889-instant-deposit

shendy-a8c committed:
Update client/components/account-balances/test/index.test.tsx

Co-authored-by: Eric Jinks <3147296+Jinksi@users.noreply.github.com>

shendy-a8c committed:
Leave InstantDepositButton for old UI and create a copy of instant deposit button for the new UI.

shendy-a8c commented:
Thank you for the review, @Jinksi. Addressed your feedback in [52f1346](https://github.com/Automattic/woocommerce-payments/pull/6004/commits/52f1346d0df9f9e93efc588ff70fcdacdaa1bbec), ie:

- Copy-pasted `deposits/instant-deposits/index.js` to `deposits/instant-deposits/button-and-modal.js`
- Leave `deposits/instant-deposits/index.js` unchanged, so no need to change unit test for the existing (old) UI.
- Use `deposits/instant-deposits/button-and-modal.js` for the new UI.

I also [commented](https://github.com/Automattic/woocommerce-payments/issues/5749#issuecomment-1510498167) in https://github.com/Automattic/woocommerce-payments/issues/5749, so we remember to remove the extraneous instant deposit component.

james-allan reviewed (approved):
Changes look good and test well for me. 

It took me a little while to jump through the required hoops to get the instant button to show, but once it did it worked well.

james-allan commented:
I'm going to merge this PR on Shendy's behalf given the team is focused on other priorities at the moment.
